### ISSUE TITLE
[react/destructuring-assignment] Relax/Optout warning when already using destructuring

### ISSUE THREAD

**Issue by mgcrea** [yannickcr/eslint-plugin-react#2153]
I often encounter the following pattern that will trigger an eslint warning:

```js
  render() {
    const {getFieldProps, getFieldError} = this.props.form;
    // [...]
  }
```

The current linting rule expects:

```js
  render() {
    const {form} = this.props;
    const {getFieldProps, getFieldError} = form;
    // [...]
  }
```

Which costs a line and is unnecessarily verbose IMO as we are already using destructuring, and `form` is a required prop.

Maybe the rule could check if the line is already using destructuring and do not warn in that case? Or add an option to relax these specific cases?

**Comment by ljharb** [comment]
The goal of the rule is to prevent ever accessing anything directly off of the props object - not merely to ensure destructuring is used. While in this case (` = this.props.form`) it's fine, if it were `= this.props.getStuff()`, that would be a problem, because `getStuff` would get access to `this.props` as its receiver (`this` value).

**Comment by luizguilhermefr** [comment]
+1

An option to ignore cases like that would be **awesome**
